libsnark/zk_proof_systems/plonk/utils.tcc

```diff
@@ -316,6 +316,72 @@ std::vector<std::vector<FieldT>> plonk_gates_matrix_transpose(
     return gates_matrix_transpose;
 }
 
+template<typename FieldT>
+void plonk_generate_constants_k1_k2(FieldT &k1, FieldT &k2)
+{
+    // n = 2^s: maximum order of the H subgroup that is power of 2
+    const size_t n = std::pow(2, FieldT::s);
+    // generator of Fr^*
+    const FieldT g = FieldT::multiplicative_generator;
+    // set k1 = g^{2s} \notin H
+    k1 = g ^ n;
+    // set k2 to a quadratic nonresidue of Fr^*
+    k2 = FieldT::nqr;
+    // assert k1,k2 are valid
+    assert(plonk_are_valid_constants_k1_k2(k1, k2));
+}
+
+template<typename FieldT>
+void plonk_generate_random_constants_k1_k2(FieldT &k1_result, FieldT &k2_result)
+{
+    // n = 2^s: maximum order of the H subgroup that is power of 2
+    const size_t n = std::pow(2, FieldT::s);
+    FieldT k1, k2;
+    // choose k1
+    do {
+        k1 = FieldT::random_element();
+    } while ((k1 ^ n) == FieldT::one());
+    // choose k2
+    FieldT k1_over_k2;
+    do {
+        k2 = FieldT::random_element();
+        k1_over_k2 = k1 * k2.inverse();
+    } while (((k2 ^ n) == FieldT::one()) ||
+             (((k1_over_k2) ^ n) == FieldT::one()));
+    k1_result = k1;
+    k2_result = k2;
+    // assert k1,k2 are valid
+    assert(plonk_are_valid_constants_k1_k2(k1, k2));
+}
+
+// The function checks if the following three conditions are
```

Comment:
I was thinking it could sit inside the function, as for other comments in the implementation.
I guess it's OK here if you prefer, as it's not a doc comment.

Reply:
Sure, I don't have a strong preference. Moved inside the function.